include/xplat/Starboard/SmartTypes.h

```diff
@@ -305,16 +305,250 @@ bool operator!=(const Any& other, const AutoId<TObj, TLifetimeTraits>& val) {
 
 #ifdef CF_EXPORT // Quick way to detect CoreFoundation.
 namespace woc {
+// Unsafe - a direct non-refcounting store of a refcountable object.
+struct CFLifetimeUnsafe {
+    inline static CFTypeRef store(CFTypeRef* destination, CFTypeRef value) {
+        *destination = value;
+        return value;
+    }
+};
+
+// Retain - a retaining/releasing store of a refcountable object.
+//  - Releases the refcounted value at the destination while never allowing
+//    *destination to hold a non-owning reference.
+struct CFLifetimeRetain {
+    inline static CFTypeRef store(CFTypeRef* destination, CFTypeRef value) {
+        if (*destination == value) {
+            return value;
+        }
+
+        if (*destination) {
+            CFRelease(static_cast<CFTypeRef>(*destination));
+        }
+
+        return *destination = ((value ? CFRetain(value) : nullptr));
+    }
+};
+
+namespace details {
+template <typename T, typename TWrapped>
+class AutoCFRef {
+private:
+    T* ptr;
+
+public:
+    AutoCFRef(T* ptr): ptr(ptr) { }
+
+    operator TWrapped*() const {
+        ptr->attach(nullptr);
+        return &ptr->get();
+    }
+
+    operator void**() const {
+        ptr->attach(nullptr);
+        return &ptr->get();
+    }
+
+    operator TWrapped() const {
+        return ptr->get();
+    }
+
+    operator void*() const {
+        return ptr->get();
+    }
+};
+}
+
+template <typename T, typename TLifetimeTraits = CFLifetimeRetain>
+class AutoCF {
+private:
+    inline CFTypeRef* _addressof() {
+        using TMutable = typename std::add_pointer<typename std::remove_const<typename std::remove_pointer<T>::type>::type>::type;
+        return const_cast<CFTypeRef*>(reinterpret_cast<void**>(const_cast<TMutable*>(&_val)));
+    }
+
+    template <typename TOther>
+    using TCanConvertFrom = std::is_convertible<TOther, T>;
+
+    template <typename TOther>
+    using TCanConvertTo = std::is_convertible<T, TOther>;
+
+    T _val = reinterpret_cast<T>(0);
+
+public:
+    AutoCF() : _val(nullptr) {
+    }
+
+    explicit AutoCF(const T& val) : _val(nullptr) {
+        TLifetimeTraits::store(_addressof(), static_cast<CFTypeRef>(val));
+    }
+
+    explicit AutoCF(T&& val): _val(val) {
+        val = nullptr;
+    }
+
+    template <typename TOtherLifetime>
+    AutoCF(const AutoCF<T, TOtherLifetime>& other)
+        : _val(nullptr) {
+        TLifetimeTraits::store(_addressof(), static_cast<CFTypeRef>(other._val));
+    }
+
+    // Copy from same lifetime.
+    AutoCF(const AutoCF<T, TLifetimeTraits>& other) : _val(nullptr) {
```

Comment:
Yes. I'm fairly certain the assignment operator explanation below applies here as well.